src/logging/quiet/run-tracker.ts

```diff
@@ -0,0 +1,652 @@
+ *    ┌────▼────┐          ┌──▼─┐
+ *    │executing├──────────►done│
+ *    └─────────┘          └────┘
+ */
+type StatusLineState =
+  /**
+   * The run has just begun.
+   */
+  | 'initial'
+  /**
+   * After analysis has started, but before it has finished.
+   */
+  | 'analyzing'
+  /**
+   * Analysis has finished, we're executing the run. We may have encountered
+   * some failures, but the run isn't over yet and we're still giving a
+   * status line with the spinner.
+   */
+  | 'executing'
+  /**
+   * As far as the spinner is concerned, we've reached a terminal state.
+   *
+   * Maybe the run succeeded. Maybe it failed. Maybe we're now executing the
+   * root script and so instead of the spinner we're purely passing through its
+   * output. In any case, no more spinner, no more status line.
+   */
+  | 'done';
+
+/**
+ * The info we expect to get from an analysis in order to report progress
+ * info.
+ */
+interface AnalysisInfo {
+  readonly rootScript: ScriptReferenceString;
+  // The scripts that we need to actually run, including services, scripts
+  // we can skip because of freshness / output restoration, etc.
+  readonly scriptsWithCommands: ReadonlySet<ScriptReferenceString>;
+  // Whether we might need to run services as part of this run.
+  readonly hasServices: boolean;
+}
+
+/**
+ * Tracks the state of a run, and produces a single line of status text.
+ *
+ * A QuietLogger usually just has one of these, but in --watch mode we use
+ * one per iteration.
+ */
+export class QuietRunLogger {
+  /**
+   * Currently running scripts, or failed scripts that we're about to
+   * report on. A script is added to this when it starts, and removed
+   * only when it successfully exits.
+   *
+   * Keyed by this._getKey
+   */
+  private readonly _running = new StackMap<
+    ScriptReferenceString,
+    ScriptState
+  >();
+  /** The number of commands we've started. */
+  private _ran = 0;
+  /**
+   * The number of scripts with commands that were either fresh or whose output
+   * we restored.
+   */
+  private _skipped = 0;
+  private _encounteredFailures = false;
+  private _servicesRunning = 0;
+  private _servicesStarted = 0;
+  private _servicesPersistedFromPreviousRun = 0;
+  private _analysisInfo: AnalysisInfo | undefined = undefined;
+  private _finishedScriptsWithCommands = new Set<ScriptReferenceString>();
+  private _statusLineState: StatusLineState = 'initial';
+  private readonly _startTime = Date.now();
+  private readonly _rootPackage: string;
+  private readonly _defaultLogger: DefaultLogger;
+  private readonly _writeoverLine;
+  /**
+   * Sometimes a script will fail multiple times, but we only want to report
+   * about the first failure for it that we find. Sometimes a script will
+   * even end up with multiple failures for the same reason, like multiple
+   * non-zero-exit-code events. This looks to be coming at the NodeJS level
+   * or above, and so we just cope.
+   *
+   * Keyed by this._getKey
+   */
```

Comment:
Is this comment for _scriptsWithAlreadyReportedErrors? Remove the blank line between them?